src/detect-engine-alert.c

```diff
@@ -188,44 +231,54 @@ int PacketAlertRemove(Packet *p, uint16_t pos)
+            PacketAlertInsertPos(det_ctx, s, p, tx_id, flags, p->alerts.cnt);
+            SCLogDebug("Added signature %" PRIu32 " to packet %" PRIu64 "", s->id, p->pcap_cnt);
+        } else {
+            /* find position to insert */
+            int16_t i = MemMovePacketAlertQueuePos(p, s);
+            PacketAlertInsertPos(det_ctx, s, p, tx_id, flags, i);
+            SCLogDebug("Added signature %" PRIu32 " to packet %" PRIu64 "", s->id, p->pcap_cnt);
+        }
 
-    if (p->alerts.cnt == PACKET_ALERT_MAX)
-        return 0;
-
-    SCLogDebug("sid %"PRIu32"", s->id);
-
-    /* It should be usually the last, so check it before iterating */
-    if (p->alerts.cnt == 0 || (p->alerts.cnt > 0 &&
-                               p->alerts.alerts[p->alerts.cnt - 1].num < s->num)) {
-        /* We just add it */
-        p->alerts.alerts[p->alerts.cnt].num = s->num;
-        p->alerts.alerts[p->alerts.cnt].action = s->action;
-        p->alerts.alerts[p->alerts.cnt].flags = flags;
-        p->alerts.alerts[p->alerts.cnt].s = s;
-        p->alerts.alerts[p->alerts.cnt].tx_id = tx_id;
-        p->alerts.alerts[p->alerts.cnt].frame_id =
-                (flags & PACKET_ALERT_FLAG_FRAME) ? det_ctx->frame_id : 0;
+        /* Update the count */
+        p->alerts.cnt++;
     } else {
-        /* We need to make room for this s->num
-         (a bit ugly with memcpy but we are planning changes here)*/
-        for (i = p->alerts.cnt - 1; i >= 0 && p->alerts.alerts[i].num > s->num; i--) {
-            memcpy(&p->alerts.alerts[i + 1], &p->alerts.alerts[i], sizeof(PacketAlert));
+        SCLogDebug("Reached packet_alert_max.");
+        /* If we reach packet_alert_max, remove lower priority
+         * rules and keep newer, higher priority ones.
+         * If Suri wants to append a signature whose priority is lower than the
+         * ones already queued and we are at packet_alert_max, it isn't queued. */
+        int16_t num_diff = s->num - p->alerts.alerts[p->alerts.cnt - 1].num;
+        if (num_diff == 0 || num_diff == -1) {
+            /* Replace last position in queue */
+            SCLogDebug("Replacing lower priority signature %" PRIu32 " with sid %" PRIu32
+                       " in packet %" PRIu64 "",
+                    p->alerts.alerts[p->alerts.cnt - 1].s->id, s->id, p->pcap_cnt);
+            PacketAlertInsertPos(det_ctx, s, p, tx_id, flags, (p->alerts.cnt - 1));
+        } else if (num_diff < -1) {
+            /* If the new signature's internal id isn't equal/adjacent to the last one from the
+             * queue, find the correct position, to keep queue sorted by rule priority */
+            int16_t i = MemMovePacketAlertQueuePos(p, s);
+            SCLogDebug("Replacing lower priority signature %" PRIu32 " with sid %" PRIu32
+                       " in packet %" PRIu64 "",
+                    p->alerts.alerts[i].s->id, s->id, p->pcap_cnt);
+            PacketAlertInsertPos(det_ctx, s, p, tx_id, flags, i);
```

Comment:
Check if this is enough to "push" lower priority alerts out of the queue or if more work is needed here.

Reply:
I think this is enough : there is no reference to dangling pointers to free

Reply:
thanks :)